core_data_modules/traced_data/io.py

```diff
@@ -314,6 +314,97 @@ def import_coda_2_to_traced_data_iterable(cls, user, data, message_id_key, schem
+
+        TODO: Data which has been assigned a code under one scheme but none of the others needs to coded as NC not NR
+        TODO: Or, do this in Coda so as to remove ambiguity from the perspective of the RAs?
+
+        :param user: Identifier of user running this program.
+        :type user: str
+        :param data: TracedData objects to be coded using the Coda file.
+        :type data: iterable of TracedData
+        :param message_id_key: Key in TracedData objects of the message ids.
+        :type message_id_key: str
+        :param scheme_keys: Dictionary of (key in TracedData objects to assign labels to) ->
+                            (Scheme in the Coda messages file to retrieve the labels from)
+        :type scheme_keys: dict of str -> Scheme
+        :type scheme_keys: dict of str -> list of str
+        :param f: Coda data file to import codes from, or None. If None, assigns NOT_REVIEWED codes to everything.
+        :type f: file-like | None
+        """
+        if f is None:
+            f = cls._make_empty_file()
+
+        # Build a lookup table of MessageID -> SchemeID -> Labels
+        coda_dataset = cls._dataset_lut_from_messages_file(f)
+
+        # Filter out TracedData objects that do not contain a message id key
+        data = [td for td in data if message_id_key in td]
+
+        # Apply the labels from Coda to each TracedData item in data
+        for td in data:
+            for coded_key, scheme in scheme_keys.items():
+                # Get all the labels assigned to this scheme across all the virtual schemes in Coda,
+                # and sort oldest first.
+                labels = []
+                all_scheme_labels = coda_dataset.get(td[message_id_key], dict())
+                for scheme_id, scheme_labels in all_scheme_labels.items():
+                    if scheme_id.startswith(scheme.scheme_id):
+                        labels.extend(scheme_labels)
+                labels.sort(key=lambda l: isoparse(l["DateTimeUTC"]))
+
+                # Get the currently assigned list of labels for this multi-coded scheme,
+                # and construct a look-up table of scheme id -> label
+                td_labels = td.get(coded_key, [])
+                td_labels_lut = {label["SchemeID"]: label for label in td_labels}
+
+                for label in labels:
+                    # Update the relevant label in this traced data's list of labels with the new label,
+                    # and append the whole new list to the traced data.
+                    td_labels_lut[label["SchemeID"]] = label
+
+                    td_labels = list(td_labels_lut.values())
+                    td.append_data({coded_key: td_labels},
+                                   Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+                # Delete any labels that are SPECIAL-MANUALLY_UNCODED
```

Comment:
Can I check what the effect of this is? 

If we manually uncode a label we should get an NC later in the pipeline, we certainly shouldn't pick up the label that was manually uncoded.

Reply:
No we shouldn't. NC is an explicitly assigned code meaning that this data couldn't be coded under the provided scheme, not that we "undid" an assigned code - this should be converted to NR, as it is in the next block